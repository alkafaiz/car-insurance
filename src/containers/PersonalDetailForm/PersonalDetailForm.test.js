import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { PersonalDetailForm } from "./index";
import React from "react";
import { subYears } from "date-fns";

afterEach(cleanup);

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: jest.fn()
  })
}));

const defaultProps = {
  nextStep: jest.fn(),
  setPersonalDetails: jest.fn(),
  initialValues: {
    firstName: "",
    lastName: "",
    dob: "",
    email: ""
  }
};

const setup = props =>
  render(<PersonalDetailForm {...defaultProps} {...props} />);

describe("PersonalDetailForm container", () => {
  it("should render", () => {
    const wrapper = setup();

    expect(wrapper.getByText("Personal detail")).toBeDefined();
    expect(wrapper.getByText("Personal detail")).toBeTruthy();
  });

  it("should render without crashing", () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("should validate required fields", async () => {
    const { getByText, getAllByText } = setup();

    const button = getByText("Next");

    fireEvent.click(button);

    const requiredLabel = await waitFor(() => getAllByText("Required"));

    expect(requiredLabel.length).toBe(4);
  });

  it("should validate email format", async () => {
    const { getByText, getByPlaceholderText } = setup();

    const emailInput = getByPlaceholderText("Email address");
    const button = getByText("Next");

    fireEvent.change(emailInput, { target: { value: "test.com" } });
    fireEvent.click(button);

    const invalidEmailText = await waitFor(() =>
      getByText("Invalid email format")
    );

    expect(invalidEmailText).toBeDefined();
  });

  it("should submit form", async () => {
    const initialValues = {
      firstName: "",
      lastName: "",
      dob: subYears(new Date(), 20),
      email: ""
    };

    const mockData = {
      firstName: "John",
      lastName: "Doe",
      email: "johndoe@gmail.com"
    };

    const { getByPlaceholderText, getByText } = setup({
      initialValues
    });

    const firstNameInput = getByPlaceholderText("First name");
    const lastNameInput = getByPlaceholderText("Last name");
    const emailInput = getByPlaceholderText("Email address");
    const button = getByText("Next");

    fireEvent.change(firstNameInput, { target: { value: mockData.firstName } });
    fireEvent.change(lastNameInput, { target: { value: mockData.lastName } });
    fireEvent.change(emailInput, { target: { value: mockData.email } });

    await waitFor(() => firstNameInput);
    await waitFor(() => lastNameInput);
    await waitFor(() => emailInput);

    expect(firstNameInput.value).toBe("John");
    expect(lastNameInput.value).toBe("Doe");
    expect(emailInput.value).toBe("johndoe@gmail.com");

    await waitFor(() => fireEvent.click(button));

    const loading = getByText("Loading...");
    expect(loading).toBeDefined();
    expect(loading).toBeTruthy();
    expect(defaultProps.setPersonalDetails).toHaveBeenCalledTimes(1);
    expect(defaultProps.setPersonalDetails).toHaveBeenCalledWith({
      ...mockData,
      dob: initialValues.dob.toISOString()
    });
  });
});
