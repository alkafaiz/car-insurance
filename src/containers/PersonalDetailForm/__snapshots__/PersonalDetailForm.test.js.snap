// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalDetailForm container should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <h2
        class="chakra-heading css-1r2fo6r"
      >
        Personal detail
      </h2>
      <form
        action="#"
      >
        <div
          class="chakra-form-control css-1lxvb1z"
          role="group"
        >
          <label
            class="chakra-form__label css-5nk1it"
            for="first-name"
            id="field-9-label"
          >
            First name
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1u8qly9"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="field-9-feedback"
            aria-required="true"
            class="chakra-input css-1u8qly9"
            id="first-name"
            name="firstName"
            placeholder="First name"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1lxvb1z"
          role="group"
        >
          <label
            class="chakra-form__label css-5nk1it"
            for="last-name"
            id="field-10-label"
          >
            Last name
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1u8qly9"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="field-10-feedback"
            aria-required="true"
            class="chakra-input css-1u8qly9"
            id="last-name"
            name="lastName"
            placeholder="Last name"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1lxvb1z"
          role="group"
        >
          <label
            class="chakra-form__label css-5nk1it"
            for="email"
            id="field-11-label"
          >
            Email
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1u8qly9"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-describedby="field-11-feedback"
            aria-required="true"
            class="chakra-input css-1u8qly9"
            id="email"
            name="email"
            placeholder="Email address"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-i55s4y"
          role="group"
        >
          <label
            class="chakra-form__label css-5nk1it"
            for="field-12"
            id="field-12-label"
          >
            Date of birth
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1u8qly9"
              role="presentation"
            >
              *
            </span>
          </label>
          <div
            class="react-datepicker-wrapper datepicker-container"
          >
            <div
              class="react-datepicker__input-container"
            >
              <input
                aria-describedby="field-12-feedback"
                aria-required="true"
                class="chakra-input css-1u8qly9"
                id="field-12"
                name="dob"
                placeholder="Date of birth"
                readonly=""
                required=""
                value=""
              />
            </div>
          </div>
        </div>
        <button
          class="chakra-button css-1yyty7m"
          type="submit"
        >
          Next
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <h2
      class="chakra-heading css-1r2fo6r"
    >
      Personal detail
    </h2>
    <form
      action="#"
    >
      <div
        class="chakra-form-control css-1lxvb1z"
        role="group"
      >
        <label
          class="chakra-form__label css-5nk1it"
          for="first-name"
          id="field-9-label"
        >
          First name
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1u8qly9"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-describedby="field-9-feedback"
          aria-required="true"
          class="chakra-input css-1u8qly9"
          id="first-name"
          name="firstName"
          placeholder="First name"
          required=""
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1lxvb1z"
        role="group"
      >
        <label
          class="chakra-form__label css-5nk1it"
          for="last-name"
          id="field-10-label"
        >
          Last name
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1u8qly9"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-describedby="field-10-feedback"
          aria-required="true"
          class="chakra-input css-1u8qly9"
          id="last-name"
          name="lastName"
          placeholder="Last name"
          required=""
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-1lxvb1z"
        role="group"
      >
        <label
          class="chakra-form__label css-5nk1it"
          for="email"
          id="field-11-label"
        >
          Email
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1u8qly9"
            role="presentation"
          >
            *
          </span>
        </label>
        <input
          aria-describedby="field-11-feedback"
          aria-required="true"
          class="chakra-input css-1u8qly9"
          id="email"
          name="email"
          placeholder="Email address"
          required=""
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-i55s4y"
        role="group"
      >
        <label
          class="chakra-form__label css-5nk1it"
          for="field-12"
          id="field-12-label"
        >
          Date of birth
          <span
            aria-hidden="true"
            class="chakra-form__required-indicator css-1u8qly9"
            role="presentation"
          >
            *
          </span>
        </label>
        <div
          class="react-datepicker-wrapper datepicker-container"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              aria-describedby="field-12-feedback"
              aria-required="true"
              class="chakra-input css-1u8qly9"
              id="field-12"
              name="dob"
              placeholder="Date of birth"
              readonly=""
              required=""
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="chakra-button css-1yyty7m"
        type="submit"
      >
        Next
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
